### PR TITLE
[web] Running integration tests on Safari on Local

### PR DESCRIPTION
**Pull request by nturgut**
Changes to felt to be able to run integration tests on Safari both on LUCI and on local MacOs machines.

fixes: https://github.com/flutter/flutter/issues/57348

currently blocked on: https://github.com/flutter/flutter/issues/57758

### CODE REVIEW

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 1]
```diff
@@ -4,57 +4,37 @@
...
+            ? ChromeDriverManager(_browser)
+            : (_browser == 'safari' && io.Platform.isMacOS)
+                ? SafariDriverManager(_browser)
+                : throw StateError(unsupportedConfigurationWarning);
 
   Future<bool> runTests() async {
-    if (_browser != 'chrome') {
-      print('WARNING: integration tests are only supported on chrome for now');
+    if (_browser != 'chrome' && _browser != 'safari' && io.Platform.isMacOS) {
+      print(unsupportedConfigurationWarning);
```
Comment (yjbanov) [inline lib/web_ui/dev/integration_tests_manager.dart:34]: Warning should go to `io.stderr`.

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 2]
```diff
@@ -219,34 +164,19 @@ class IntegrationTestsManager {
...
-        statementToRun = '$statementToRun --chrome-binary='
-            '${preinstalledChromeExecutable()}';
-      }
-      io.stderr
-          .writeln('ERROR: Failed to run test. Exited with exit code $exitCode'
-              '. Statement to run $testName locally use the following '
-              'command:\n\n$statementToRun');
+      io.stderr.writeln(
+          'ERROR: Failed to run test. Exited with exit code $exitCode'
+          '. Statement to run $testName locally use the following '
```
Comment (yjbanov) [inline lib/web_ui/dev/integration_tests_manager.dart:178]: Remove "Statement"?

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 3]
```diff
@@ -356,6 +286,77 @@ class IntegrationTestsManager {
...
+    } else if (browser == 'safari' && io.Platform.isMacOS) {
+      return SafariIntegrationArguments();
+    } else {
+      throw StateError(unsupportedConfigurationWarning);
+    }
+  }
+
+  List<String> getTestArguments(String testName, String mode);
+
+  String getStatementToRun(String testName, String mode);
```
Comment (yjbanov) [inline lib/web_ui/dev/integration_tests_manager.dart:306]: s/Statement/Command/ ?

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 4]
```diff
@@ -356,6 +286,77 @@ class IntegrationTestsManager {
...
+  List<String> getTestArguments(String testName, String mode) {
+    return <String>[
+      'drive',
+      '--target=test_driver/${testName}',
+      '-d',
+      'web-server',
+      '--$mode',
+      '--browser-name=chrome',
+      if (isLuci) '--chrome-binary=${preinstalledChromeExecutable()}',
+      if (isLuci) '--headless',
```
Comment (yjbanov) [inline lib/web_ui/dev/integration_tests_manager.dart:320]: We should always run in headless mode, LUCI and local. The fewer special-cases we have for LUCI the more likely you are to reproduce a breakage that happens on LUCI.

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 5]
```diff
@@ -356,6 +286,77 @@ class IntegrationTestsManager {
...
+      '--local-engine=host_debug_unopt',
+    ];
+  }
+
+  String getStatementToRun(String testName, String mode) {
```
Comment (yjbanov) [inline lib/web_ui/dev/integration_tests_manager.dart:325]: This should be just `=> 'flutter ${getTestArguments(testName, mode).join(' ')}'`

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 6]
```diff
@@ -4,57 +4,37 @@
...
 
 import 'chrome_installer.dart';
+import 'driver_manager.dart';
 import 'environment.dart';
 import 'exceptions.dart';
 import 'common.dart';
 import 'utils.dart';
 
+const String unsupportedConfigurationWarning = 'WARNING: integration '
+    'tests are only supported on Chrome or on Safari(running on MacOS)';
```
Comment (yjbanov) [inline lib/web_ui/dev/integration_tests_manager.dart:16]: nit: "on Chrome and on Safari"
nit: add one space between "Safari" and "("

File: lib/web_ui/dev/driver_manager.dart [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    final SafariDriverRunner safariDriverRunner = SafariDriverRunner();
+
+    final io.Process process =
+        await safariDriverRunner.runDriver(version: 'system');
+
+    processesToCleanUp.add(process);
+  }
+}
+
+/// Interface for preparing the browser driver before running the integration
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart:85]: done.

File: lib/web_ui/dev/driver_manager.dart [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+abstract class DriverManager {
+  /// Installation directory for browser's driver.
+  ///
+  /// Always re-install since driver can change frequently.
+  /// It usually changes with each the browser version changes.
+  /// A better solution would be installing the browser and the driver at the
+  /// same time.
+  // TODO(nurhan): https://github.com/flutter/flutter/issues/53179. Partly
+  // solved. Remaining local integration tests using the locked Chrome version.
+  final io.Directory _browserDriverDir;
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart:96]: Done. (Made it public instead) 
It says: "Undefined name 'protected' used as an annotation." There are no `@protected` on fields under flutter/engine.

File: lib/web_ui/dev/driver_manager.dart [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+  /// This is the parent directory for all drivers.
+  ///
+  /// This directory is saved to [temporaryDirectories] and deleted before
+  /// tests shutdown.
+  final io.Directory _drivers;
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart:102]: done. (to public) 

same comment applies.

File: lib/web_ui/dev/driver_manager.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart:96]: ah never mind it is under meta package :)

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 10]
```diff
@@ -4,57 +4,37 @@
...
-import 'package:web_driver_installer/chrome_driver_installer.dart';
 
 import 'chrome_installer.dart';
+import 'driver_manager.dart';
 import 'environment.dart';
 import 'exceptions.dart';
 import 'common.dart';
 import 'utils.dart';
 
+const String unsupportedConfigurationWarning = 'WARNING: integration '
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:15]: done.

File: lib/web_ui/dev/integration_tests_manager.dart (hunk 6 above)
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:16]: done

File: lib/web_ui/dev/integration_tests_manager.dart (hunk 2 above)
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:178]: done.

File: lib/web_ui/dev/integration_tests_manager.dart (hunk 3 above)
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:306]: done.

File: lib/web_ui/dev/integration_tests_manager.dart (hunk 5 above)
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:325]: But it differs depending on where it runs. Check next lines. There is an if statement for LUCI. Trying to squeeze this method like that will make it less readable.

Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:325]: changed the one in the other class (SafariIntegrationArguments).

File: lib/web_ui/dev/integration_tests_manager.dart (hunk 1 above)
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:34]: done.

File: lib/web_ui/dev/driver_manager.dart [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+      await _verifyDriverForLUCI();
+    }
+    await _startDriver(_browserDriverDir.path);
+  }
+
+  Future<void> _installDriver();
+
+  void _verifyDriverForLUCI();
+
+  Future<void> _startDriver(String driverPath);
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart:127]: done.

File: lib/web_ui/dev/integration_tests_manager.dart (hunk 4 above)
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:320]: Good catch, thanks.

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 12]
```diff
@@ -4,57 +4,37 @@
...
+      : _driverManager = _browser == 'chrome'
+            ? ChromeDriverManager(_browser)
+            : (_browser == 'safari' && io.Platform.isMacOS)
+                ? SafariDriverManager(_browser)
+                : throw StateError(unsupportedConfigurationWarning);
 
   Future<bool> runTests() async {
-    if (_browser != 'chrome') {
-      print('WARNING: integration tests are only supported on chrome for now');
+    if (_browser != 'chrome' && _browser != 'safari' && io.Platform.isMacOS) {
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart:35]: done.

File: lib/web_ui/dev/integration_tests_manager.dart [hunk 13]
```diff
@@ -4,57 +4,37 @@
...
-            'drivers',
-            _browser,
-            '${_browser}driver-${io.Platform.operatingSystem.toString()}')),
-        this._drivers = io.Directory(
-            pathlib.join(environment.webUiDartToolDir.path, 'drivers'));
+      : _driverManager = _browser == 'chrome'
+            ? ChromeDriverManager(_browser)
+            : (_browser == 'safari' && io.Platform.isMacOS)
+                ? SafariDriverManager(_browser)
+                : throw StateError(unsupportedConfigurationWarning);
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/dev/integration_tests_manager.dart]: So true, good suggestion, thanks!

File: lib/web_ui/dev/driver_manager.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/dev/driver_manager.dart:100]: `@protected` doesn't make sense if the field is private. It should be public _and_ `@protected`.

File: lib/web_ui/dev/driver_manager.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/dev/driver_manager.dart:107]: ditto

File: lib/web_ui/dev/driver_manager.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/dev/driver_manager.dart:133]: ditto

File: lib/web_ui/dev/driver_manager.dart [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+  Future<void> _installDriver() async {
+    if (_browserDriverDir.existsSync()) {
+      _browserDriverDir.deleteSync(recursive: true);
+    }
+
+    _browserDriverDir.createSync(recursive: true);
+    temporaryDirectories.add(_drivers);
+
+    io.Directory temp = io.Directory.current;
```
Comment (yjbanov) [inline lib/web_ui/dev/driver_manager.dart:31]: This can be `final`.

File: lib/web_ui/dev/driver_manager.dart [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    io.Directory.current = _browserDriverDir;
+
+    // TODO(nurhan): https://github.com/flutter/flutter/issues/53179
+    final String chromeDriverVersion = await queryChromeDriverVersion();
+    ChromeDriverInstaller chromeDriverInstaller =
+        ChromeDriverInstaller.withVersion(chromeDriverVersion);
+    await chromeDriverInstaller.install(alwaysInstall: true);
+    io.Directory.current = temp;
```
Comment (yjbanov) [inline lib/web_ui/dev/driver_manager.dart:39]: If a statement between line 32 and this line throws an exception, the current directory will be left pointing to `_browserDriverDir`. Consider wrapping the code in `try/finally` and restoring the current directory in the `finally` block. However, an even better design would be to not rely on mutable global state. For example, is it possible to pass the current directory to whoever needs it?

File: lib/web_ui/dev/driver_manager.dart [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  /// Driver should already exist on LUCI as a CIPD package.
+  ///
+  /// Throw an error if directory does not exists.
```
Comment (yjbanov) [inline lib/web_ui/dev/driver_manager.dart:44]: dartdocs should start with the sentence that explains what the function does, so I'd swap the two paragraphs.

https://dart.dev/guides/language/effective-dart/documentation#prefer-starting-function-or-method-comments-with-third-person-verbs

File: lib/web_ui/dev/driver_manager.dart [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+/// Abstract class for preparing the browser driver before running the integration
+/// tests.
+abstract class DriverManager {
+  /// Installation directory for browser's driver.
+  ///
+  /// Always re-install since driver can change frequently.
+  /// It usually changes with each the browser version changes.
+  /// A better solution would be installing the browser and the driver at the
+  /// same time.
```
Comment (yjbanov) [inline lib/web_ui/dev/driver_manager.dart:96]: Should the second paragraph be on another method, e.g. `_installDriver`, or perhaps the class itself? I doubt I'd be looking for information about how often we install the driver on a field called `_browserDriverDir`.

File: lib/web_ui/dev/driver_manager.dart (hunk 14 above)
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart]: done

File: lib/web_ui/dev/driver_manager.dart (hunk 16 above)
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart]: done

File: lib/web_ui/dev/driver_manager.dart (hunk 17 above)
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart]: done

File: lib/web_ui/dev/driver_manager.dart (hunk 15 above)
Comment (nturgut) [reply lib/web_ui/dev/driver_manager.dart]: I did the first one, the second one is a good suggestion yet it requires me to change the other repo.

**Comment by nturgut** [comment]
Merging the PR the error in the MaciOSEngine logs' step is not relevant to this PR's changes (This PR code path is Web Engine Test Harness)

**Pull request merged**